src/client/components/scroller/scroller.tsx

```diff
@@ -2,44 +2,380 @@ require('./scroller.css');
+
+    if (side === 'top') return scrollTop > 0;
+    if (side === 'left') return scrollLeft > 0;
+    if (side === 'bottom') return layout.bodyHeight - scrollTop > viewportHeight;
+    if (side === 'right') return layout.bodyWidth - scrollLeft > viewportWidth;
+
+    throw new Error('Unknow side for shadow : ' + side);
+  }
+
+  renderShadow(side: string): JSX.Element {
+    if (['top', 'bottom', 'left', 'right'].indexOf(side) === -1) {
+      throw new Error('Unknown shadow : ' + side);
+    }
+
+    if (!(this.props.layout as any)[side]) return null; // no gutter ? no shadow.
+    if (!this.shouldHaveShadow(side)) return null;
+
+    const name = `${side}Shadow`;
+
+    var styler = (this as any)[`get${this.firstUp(name)}Style`].bind(this);
+
+    return <div className={this.toKebabCase(name)} style={styler()}/>;
+  }
+
+  renderCorner(side: string): JSX.Element {
+    if (['topLeft', 'topRight', 'bottomLeft', 'bottomRight'].indexOf(side) === -1) {
+      throw new Error('Unknown corner : ' + side);
+    }
+
+    return this.renderTile(side + 'Corner');
+  }
+
+  componentDidUpdate() {
+    const rect = this.getDOMElement('Scroller').getBoundingClientRect();
+    const { top, right, bottom, left } = this.props.layout;
+
+    const newHeight = rect.height - top - bottom;
+    const newWidth = rect.width - left - right;
+
+    if (this.state.viewportHeight !== newHeight || this.state.viewportWidth !== newWidth) {
+      this.setState({viewportHeight: newHeight, viewportWidth: newWidth});
+    }
+  }
+
   render() {
-    const { style, onScroll, onMouseLeave, onMouseMove, onClick } = this.props;
-    return <div
-      className="scroller"
-      ref="base"
-      onScroll={onScroll}
-      onMouseLeave={onMouseLeave || null}
-      onMouseMove={onMouseMove || null}
-      onClick={onClick || null}
-    >
-      <div className="scroller-inner" style={style}></div>
+    const { body, bodyOverlay, onMouseLeave } = this.props;
+
+    return <div className="scroller" ref="Scroller">
+
+      <div className="body" style={this.getBodyStyle()}>{body}</div>
+
+      {this.renderGutter("top")}
+      {this.renderGutter("right")}
+      {this.renderGutter("bottom")}
+      {this.renderGutter("left")}
+
+      {this.renderCorner("topLeft")}
+      {this.renderCorner("topRight")}
+      {this.renderCorner("bottomLeft")}
+      {this.renderCorner("bottomRight")}
+
+      {this.renderShadow("top")}
+      {this.renderShadow("right")}
+      {this.renderShadow("bottom")}
+      {this.renderShadow("left")}
+
+      <div className="body-overlay" style={this.getBodyOverlayStyle()}>{bodyOverlay}</div>
+
+      <div
+        className="event-container"
+        ref="eventContainer"
+        onScroll={this.onScroll.bind(this)}
+        onClick={this.onClick.bind(this)}
+        onMouseMove={this.onMouseMove.bind(this)}
+        onMouseLeave={onMouseLeave || null}
+       >
+        <div className="event-target" style={this.getTargetStyle()}/>
```

Comment:
Can we name is something better? Scrolling expander? Jaws of life?

http://s1.cdn.autoevolution.com/images/news/app-to-guide-the-jaws-of-life-33667_1.jpg